### PR TITLE
feat: Support *, /, - operations between two DiffractionObjects or scalar 

### LINKED ISSUE #187: write tests for...
### Problem
- [ ] __eq__
- [ ] __add__
- [ ] __radd__
- [ ] __sub__
- [ ] __rsub__
- [ ] __mul__
- [ ] __rmul__
- [ ] __truediv__
- [ ] __rtruediv__

before doing this talk to @yucongalicechen or @sbillinge about new data structure for the arrays in the DOs.

### PR DESCRIPTION
**Pull request by bobleesj**
Closes #187 - 

No docstrings made at the moment, but I will do in the next PR working on the issue https://github.com/diffpy/diffpy.utils/issues/239

### CODE REVIEW

File: src/diffpy/utils/diffraction_objects.py [hunk 1]
```diff
@@ -217,95 +217,48 @@ def __add__(self, other):
...
+    __rtruediv__ = __truediv__
+
     def _check_operation_compatibility(self, other):
         if not isinstance(other, (DiffractionObject, int, float)):
             raise TypeError(invalid_add_type_emsg)
         if isinstance(other, DiffractionObject):
             self_yarray = self.all_arrays[:, 0]
             other_yarray = other.all_arrays[:, 0]
-            if len(self_yarray) != len(other_yarray):
+            if self_yarray.shape != other_yarray.shape:
```
Comment (bobleesj) [inline src/diffpy/utils/diffraction_objects.py:259]: @sbillinge as suggested - using `shape`

File: tests/test_diffraction_objects.py [hunk 2]
```diff
@@ -713,75 +713,176 @@ def test_copy_object(do_minimal):
 
 
 @pytest.mark.parametrize(
-    "starting_all_arrays, scalar_to_add, expected_all_arrays",
+    "operation, starting_yarray, scalar_value, expected_yarray",
```
Comment (bobleesj) [inline tests/test_diffraction_objects.py:716]: 1.

operation b/w scalar and do

File: tests/test_diffraction_objects.py [hunk 3]
```diff
@@ -713,75 +713,176 @@ def test_copy_object(do_minimal):
...
-    "expected_do_2_all_arrays_with_y_summed",
+    "operation, " "expected_do_1_all_arrays_with_y_modified, " "expected_do_2_all_arrays_with_y_modified",
     [
-        # Test addition of two DO objects, expect combined yarray values and no change to xarrays ((q, tth, d)
-        (  # C1: Add two DO objects, expect sum of yarray values
-            (np.array([[1.0, 0.51763809, 30.0, 12.13818192], [2.0, 1.0, 60.0, 6.28318531]]),),
-            (np.array([[1.0, 6.28318531, 100.70777771, 1], [2.0, 3.14159265, 45.28748053, 2.0]]),),
-            (np.array([[2.0, 0.51763809, 30.0, 12.13818192], [4.0, 1.0, 60.0, 6.28318531]]),),
-            (np.array([[2.0, 6.28318531, 100.70777771, 1], [4.0, 3.14159265, 45.28748053, 2.0]]),),
+        # Test addition, subtraction, multiplication, and division of two DO objects
```
Comment (bobleesj) [inline tests/test_diffraction_objects.py:800]: 2.

operation b/w do1 and do2

File: tests/test_diffraction_objects.py [hunk 4]
```diff
@@ -713,75 +713,176 @@ def test_copy_object(do_minimal):
...
+    elif operation == "div":
+        do_1_y_modified = do_1 / do_2
+        do_2_y_modified = do_2 / do_1
+
+    assert np.allclose(do_1_y_modified.all_arrays, expected_do_1_all_arrays_with_y_modified)
+    assert np.allclose(do_2_y_modified.all_arrays, expected_do_2_all_arrays_with_y_modified)
+
+
+def test_operator_invalid_type(do_minimal_tth, invalid_add_type_error_msg):
+    # Add a string to a DiffractionObject, expect TypeError
```
Comment (bobleesj) [inline tests/test_diffraction_objects.py:857]: invalid case - add string

File: tests/test_diffraction_objects.py [hunk 5]
```diff
@@ -713,75 +713,176 @@ def test_copy_object(do_minimal):
...
+    for operation in operations:
+        with pytest.raises(TypeError, match=re.escape(invalid_add_type_error_msg)):
+            operation(do, invalid_value)
+        with pytest.raises(TypeError, match=re.escape(invalid_add_type_error_msg)):
+            operation(invalid_value, do)
+
+
+@pytest.mark.parametrize("operation", ["add", "sub", "mul", "div"])
+def test_operator_invalid_yarray_length(operation, do_minimal, do_minimal_tth, y_grid_size_mismatch_error_msg):
+    # Add two DO objects with different yarray lengths, expect ValueError
```
Comment (bobleesj) [inline tests/test_diffraction_objects.py:875]: invalid case - add different lengths of yarray between 2 DOs

**Comment by bobleesj** [comment]
@sbillinge ready for review!

**Comment by yucongalicechen** [comment]
Do we want to add both metadata to the resulting diffraction object?
e.g. DO3 = DO1 / DO2, and DO3 contains metadata from both objects

File: src/diffpy/utils/diffraction_objects.py (hunk 1 above)
Comment (sbillinge) [reply src/diffpy/utils/diffraction_objects.py]: the idea for using shape was to shorten code and make it more readable, so 
```
            self_yarray = self.all_arrays[:, 0]
            other_yarray = other.all_arrays[:, 0]
            if len(self_yarray) != len(other_yarray):
                raise ValueError(y_grid_length_mismatch_emsg)
```
is replaced by 
```
            if shape(self.all_arrays) != shape(other.all_arrays):
                raise ValueError(y_grid_length_mismatch_emsg)
```

**Review by sbillinge — COMMENTED** [review]
Looks good.  I tried to shorten the all_arrays shape check part of the code before merging

**Comment by bobleesj** [comment]
@sbillinge ah yes. Great! Thanks

**Pull request merged**